keanu-project/src/main/java/io/improbable/keanu/algorithms/sampling/Forward.java

```diff
@@ -0,0 +1,110 @@
+import java.util.stream.Collectors;
+
+public class Forward implements PosteriorSamplingAlgorithm {
+
+    private final KeanuRandom random;
+
+    public Forward() {
+        this(KeanuRandom.getDefaultRandom());
+    }
+
+    public Forward(KeanuRandom random) {
+        this.random = random;
+    }
+
+    /**
+     * Samples from the prior of a Probabilistic Model.
+     * Bayesian Network is the only supported implementation.
+     *
+     * Samples are taken by sampling from the prior of the desired variables in topological order.
+     *
+     * @param model the model to sample from
+     * @param variablesToSampleFrom the variables to sample from
+     * @param sampleCount the number of samples to take
+     * @return sampling samples of a computable graph
+     */
+    @Override
+    public NetworkSamples getPosteriorSamples(ProbabilisticModel model, List<? extends Variable> variablesToSampleFrom, int sampleCount) {
+        return generatePosteriorSamples(model, variablesToSampleFrom)
+            .generate(sampleCount);
+    }
+
+    @Override
+    public NetworkSamplesGenerator generatePosteriorSamples(ProbabilisticModel model, List<? extends Variable> variablesToSampleFrom) {
+        return new NetworkSamplesGenerator(setupSampler(model, variablesToSampleFrom), StatusBar::new);
+    }
+
+    private SamplingAlgorithm setupSampler(ProbabilisticModel model, List<? extends Variable> variablesToSampleFrom) {
+        List<? extends Variable> latentVariables = model.getLatentVariables();
+        Preconditions.checkArgument(latentVariables.size() > 0, "Your model must contain latent variables in order to forward sample.");
+
+        Variable latent = latentVariables.get(0);
+        Preconditions.checkArgument(latent instanceof Vertex, "The Forward Sampler only works for Variables of type Vertex.");
+
+        BayesianNetwork network = new BayesianNetwork(((Vertex) latent).getConnectedGraph());
+
+        List<Vertex> observedVertices = network.getObservedVertices();
+        checkUpstreamOfObservedDoesNotContainProbabilistic(observedVertices);
+
+        List<Vertex> verticesToSampleFrom = variablesToSampleFrom.stream().map(v -> (Vertex) v).collect(Collectors.toList());
+        Set<Vertex> allUpstreamVertices = allUpstreamVertices(verticesToSampleFrom);
+
+        List<Vertex> sortedVertices = TopologicalSort.sort(allUpstreamVertices);
+        sortedVertices = removeNonProbabilisticVerticesBeforeTheFirstProbabilistic(sortedVertices);
+
+        return new ForwardSampler(verticesToSampleFrom, sortedVertices, random);
+    }
+
+    private List<Vertex> removeNonProbabilisticVerticesBeforeTheFirstProbabilistic(List<Vertex> topologicalVertices) {
+        List<Vertex> copy = new ArrayList<>(topologicalVertices);
+
+        for (Vertex vertex : topologicalVertices) {
+            if (vertex instanceof Probabilistic) {
+                break;
+            } else {
+                copy.remove(vertex);
+            }
+        }
```

Comment:
This could be pretty expensive again.  Couple of thoughts:
1) Removing elements from the start of an ArrayList is expensive as every item has to be copied over to the left - LinkedList or queue might be better if we want to handle this in this way.
2) The .remove(Object) version we're using here will be slower than just passing the index value directly.  Given we're just popping stuff off the front remove(0) would do here?

Reply:
This will return a LinkedList  as a List for later traversal btw. This could be a problem generally but especially if some other code starts randomly accessing elements.